[Distributed] Implement distributed computed properties via special accessor

xedin opened the pull request:
This is an alternative implementation of distributed computed properties that introduces
a new `DistributedThunk` accessor kind, new access strategy, and semantics.

A reference to a getter of a distributed computed property is transparently replaced by
its thunk during SILGen (the semantics are the same for both distributed methods and
distributed computed properties).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

rdar://91283164

ktoso committed:
[Distributed] Skeleton implementation of distributed computed properties

xedin committed:
[Distributed] Synthesize thunks for distributed computed properties

xedin committed:
[AST] Add new member access semantics - DistributedThunk

`DistributedThunk` is to be used while accessing 'distributed'
computed property outside of its actor context.

xedin committed:
[Distributed] Add a new access strategy - DispatchToDistributedThunk

This strategy is used to dispatch accesses to 'distributed' computed
property to distributed thunk accessor instead of a regular getter
when access happen outside actor isolation context.

xedin committed:
[Distributed] Remove commented out code and print statements

xedin committed:
[Distributed] SILGen: Remove `isDistributed` flags from accessor code

xedin committed:
[TypeChecker] Distributed: Verify properties before attempting to synthesize a thunk

ktoso committed:
fix checkDistributedTargetResultType to also return when not diagnosing

xedin committed:
[Distributed] NFC: remove a couple of unused variables from sythensis code

xedin commented:
@DougGregor and @ktoso this is still a draft but it's functional. I'd probably need to cherry-pick some of the changes from original implementation to make it complete. Please let me know what you think.

xedin committed:
[Distributed] Decl: Add a new distributed-thunk bit

The flag is used to distinguish between regular functions/accessors
and synthesized distributed thunks.

xedin committed:
[Distributed] Synthesize 'distributed thunk' accessor for distributed computed properties

xedin committed:
[MiscDiagnostics] Don't diagnose recursive property access in distributed thunks

Distributed thunks have to refer to the "local" version
of the property directly with implicit `self.` base.

xedin committed:
[Distributed] Teach `HasIsolatedSelfRequest` about distributed thunk accessors

xedin committed:
[Distributed] NFC: XFAIL a couple of incorrect attr use tests

ktoso commented on include/swift/AST/StorageImpl.h:
```diff
@@ -110,6 +110,10 @@ class AccessStrategy {
+
+    /// The access is to a computed distributed property, and thus the
+    /// get-accessor is a distributed thunk which may perform a remote call.
+    DispatchToDistributedThunk,
```
👍

ktoso commented on lib/SILGen/SILGenLValue.cpp:
```diff
@@ -2639,6 +2647,8 @@ namespace {
         AccessKind(accessKind) {}
 
     void emitUsingStrategy(AccessStrategy strategy) {
+      auto var = dyn_cast<VarDecl>(Storage);
```
doesn't seem used?

xedin replied:
Ah, I might have not cherry-picked NFC commit, thank you!

xedin replied:
Removed!

ktoso commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -782,10 +833,24 @@ addDistributedActorCodableConformance(
+      return nullptr;
+
+    if (checkDistributedActorProperty(var, /*diagnose=*/false))
+      return nullptr;
```
👍

ktoso reviewed (approved):
Looks good to me! Though needs a @DougGregor review as well to see if it looks good now :)

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test Linux platform

ktoso commented:
Cherry picked it over: https://github.com/apple/swift/pull/59774

DougGregor commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -2309,20 +2310,46 @@ namespace {
-          .fixItInsert(decl->getAttributeInsertionLoc(true), "distributed ");
+      // Cannot reference subscripts, or stored properties.
+      auto var = dyn_cast<VarDecl>(decl);
+      if (isa<SubscriptDecl>(decl) || var) {
```
The rewritten logic here silently accepts declaration kinds it doesn't understand. I'm not sure if I can trigger a bug based on this, but the prior formulation (`!func || !func->isDistributed()`) was more robust. Can we use more straightforward logic here like:

```
if (auto func = dyn_cast<AbstractFunctionDecl>(decl)) {
 handle funcs and return
}

if (auto var = dyn_cast<VarDecl>(decl)) {
  handle vars and return
}

diagnose and return None
```

DougGregor reviewed (commented):
I'm concerned about the mangling, but the rest is polish.

xedin commented on include/swift/AST/Attr.def:
```diff
@@ -735,6 +735,12 @@ CONTEXTUAL_SIMPLE_DECL_ATTR(_local, KnownToBeLocal,
   APIBreakingToAdd | APIBreakingToRemove,
   130)
 
+SIMPLE_DECL_ATTR(_distributed_thunk, DistributedThunk,
```
Yes, cannot be spelled. It could be an attribute on the decl, wasn't sure which one is better.

xedin replied:
Switched to use a bit in the declaration and changed serialization accordingly.

ktoso replied:
Thanks!

xedin commented on lib/AST/ASTMangler.cpp:
```diff
@@ -3192,6 +3192,14 @@ void ASTMangler::appendEntity(const ValueDecl *decl) {
+      appendContextOf(decl);
+      appendDeclName(accessor->getStorage());
+      appendDeclType(accessor, FunctionMangling);
+      appendOperator("F");
```
Distributed thunk accessors are mangled the same way as distributed funcs because runtime needs a function type, if we want to add a new mangling for accessor thunks we'd need to rework a runtime aspects of remote calls. I'm going to add a comment here explaining this.

xedin replied:
Added a comment about this.

xedin committed:
[Distributed] Mangling: Add a comment about distributed accessor thunks

xedin committed:
[Distributed] Allow requesting distributed thunks on `AbstractStorageDecl`

One step towards future distributed subscripts.

xedin committed:
[TypeChecker] Distributed: rework `checkDistributedAccess`

Only distributed functions and computed properties should be
accepted, everything else is invalid.

xedin commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -431,10 +431,19 @@ static bool varIsSafeAcrossActors(const ModuleDecl *fromModule,
+        // which are allowed, but subject to the same implicit throws/async
+        // as all other distributed function declarations.
+        if (var->isDistributed()) {
+          return true;
```
@ktoso I removed the check because as Doug mentioned it doesn't seem correct.

ktoso replied:
Right that seemed wrong -- thank you!

xedin commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -1237,17 +1237,20 @@ class GetDistributedRemoteCallArgumentInitFunctionRequest :
+          GetDistributedThunkRequest,
+          FuncDecl *(llvm::PointerUnion<VarDecl *, AbstractFunctionDecl *>),
+          RequestFlags::Cached> {
+  using Originator = llvm::PointerUnion<VarDecl *, AbstractFunctionDecl *>;
```
Generalized `Originator` to accept `AbstractStorageDecl` and `AbstractFunctionDecl`.

xedin commented:
@swift-ci please test

DougGregor reviewed (approved):
Looks good, thank you!

xedin committed:
[Distributed] Generate thunk for accessor as a regular method

It used to be an accessor but that is not required because
SILDeclRef controls mangling which is the most imprortant
and could be used to emit the right reference.

xedin commented:
@swift-ci please test

ktoso commented on lib/AST/ASTMangler.cpp:
````diff
@@ -3506,5 +3491,28 @@ std::string ASTMangler::mangleDistributedThunk(const AbstractFunctionDecl *thunk
+    // appendOperator("F");
+    // appendSymbolKind(SymbolKind::DistributedThunk);
+    // return finalize();
+    // ```
````
thanks for the TODO, that's on me after we fixed the SILDeclRefs

xedin replied:
Yeah, no worries!

xedin commented:
@swift-ci please test

ktoso commented:
I'll cherry pick it over to 5.7 👍

xedin commented:
@swift-ci please test Windows platform
